### PR TITLE
Draw Energy From Users Based off of time of day

### LINKED ISSUE #83: Fix charging small batteries
- [x] Small batteries are not being charged. Look at this and fix it

- There was no bug in backend. Front end implementation

- [x] check implementation on the frontend

### PR DESCRIPTION
**Pull request by Lev-Stambler**
- Add draw energy from energy users based off of time of day
- Add date tracking into BioGrid and Bio Brain
- Add date to take action
- Add test to ensure that energy is drawn

### CODE REVIEW

**Review by RolandNaijuka — APPROVED** [review]
It is looking great. Thanks for that addition

File: biogrid/apps/api/src/app/services/Biogrid.ts [hunk 1]
```diff
@@ -72,6 +72,7 @@ export async function simulateNewBiogrid(
   const initState = biogrid.getSystemState();
   const statesJson = [biogrid.getJsonGraphDetails()];
   for (let i = 0; i < constants.simulation.NUMBER_OF_SIM_CYCLES; i++) {
+    biogrid.drainEnergyUsers()
```
Comment (aosmann7) [inline biogrid/apps/api/src/app/services/Biogrid.ts]: semicolon here.

File: biogrid/libs/biogrid-simulator/src/lib/building/building.ts [hunk 2]
```diff
@@ -93,4 +102,8 @@ export class Building implements EnergyUser {
       this.energyInJoules -= energy;
     }
   }
+
+  private getAverageEnergyUsagePerDay(hourOfDay: number): Energy {
+    return BUILDING.ENERGY_USAGE_KILOJOULE_BY_TIME_OF_DAY[hourOfDay.toString()]
```
Comment (aosmann7) [inline biogrid/libs/biogrid-simulator/src/lib/building/building.ts]: semicolon?

File: biogrid/libs/grid-simulator/src/lib/energyuser/energyuser.ts [hunk 3]
```diff
@@ -1,8 +1,13 @@
-import { ItemPosition } from '../measurements';
+import { ItemPosition, Energy } from '../measurements';
 import { GridItem } from '../grid-item';
 
+export interface EnergyUsageByTimeOfDay {
+  [hourOfDay: string]: Energy
```
Comment (aosmann7) [inline biogrid/libs/grid-simulator/src/lib/energyuser/energyuser.ts:5]: semicolon

**Review by aosmann7 — APPROVED** [review]
LGTM. Just some minor fixes

File: biogrid/apps/api/src/app/services/Biogrid.ts [hunk 4]
```diff
@@ -71,8 +71,12 @@ export async function simulateNewBiogrid(
   const biobrain = BioBrain.Instance;
   const initState = biogrid.getSystemState();
   const statesJson = [biogrid.getJsonGraphDetails()];
+  const currentDate = body.startDate;
   for (let i = 0; i < constants.simulation.NUMBER_OF_SIM_CYCLES; i++) {
```
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.ts]: can we clarify the naming of NUMBER_OF_SIM_CYCLES and change it to NUMBER_OF_SIM_HOURS instead?

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts [hunk 5]
```diff
@@ -71,8 +71,8 @@ export class SolarPanel extends EnergySource {
    * This method returns the current power that can be generated by
    * the solar panel at that given time
    */
-  async getEnergyInJoules(): Promise<Power> {
-    return await this.getPowerAmount(this.date);
+  async getEnergyInJoules(date?: Date): Promise<Power> {
+    return await this.getPowerAmount(this.date || date);
```
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts]: what if the date param is empty/null? 
Is this.date supposed to provide a default value fallback? If that's the case - do you mean `date || this.date` instead?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts [hunk 6]
```diff
@@ -23,21 +23,25 @@ import {
...
   numberOfLargeBatteryCells: number;
   numberOfSolarPanels: number;
+  startDate?: Date;
 }
 
 export class Biogrid implements Grid {
   // TODO create a singleton for the Biogrid not BiogridState
   private state: BiogridState;
 
+  private startDate: Date;
```
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:45]: nit: please comment what this startDate represents or how it is used

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts [hunk 7]
```diff
@@ -144,15 +151,27 @@ export class Biogrid implements Grid {
...
     );
   }
+
+  /**
+   * Drain the energy users according to the time of day
+   */
+  drainEnergyUsers(date: Date) {
+    this.town.getEnergyUsers().forEach((energyUser) => {
+      energyUser.decreaseEnergyAccordingToTimeOfDay(date);
+    });
```
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:174]: I think this naming is pretty confusing. Energy users are things like houses right? So you're not really "draining" them of energy, but rather providing them with energy that they need. Can we rename?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts [hunk 8]
```diff
@@ -21,35 +22,39 @@ beforeAll(() => {
...
       x: 9,
       y: 9,
-      gridItemName: name5
+      gridItemName: name5,
     } as BuildingParams),
   ];
   grid = new Biogrid(
-    new RuralArea(ruralArea, /* townWidth = */ townWidth, /* townHeight = */ townHeight),
+    new RuralArea(
+      ruralArea,
```
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts]: it's a bit confusing to have a RuralArea constructor take in something named ruralArea that is not itself a ruralArea object. 
Can we rename `ruralArea` object to be `buildingsInRuralArea` or `buildingList`

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts [hunk 9]
```diff
@@ -119,14 +149,19 @@ describe('classes', () => {
     // There are two non full buildings, and the battery has to be refilled however
     // the components are placed on the grid randomly thus we cannot guarantee the battery is refilled
     // Energy may come from the solar panels but two buildings must be refiled
-    expect(Object.keys(action.getSupplyingPaths()).length).toBeGreaterThanOrEqual(2);
+    expect(
+      Object.keys(action.getSupplyingPaths()).length
+    ).toBeGreaterThanOrEqual(2);
```
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts]: can we get an exact expectation here?

**Review by vitashubin — CHANGES_REQUESTED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts (hunk 5 above)
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts]: Yeah, I fixed this is in a future PR after figuring that all out

Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts]: In PR #85

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts (hunk 7 above)
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:174]: Oh it's more like updating energy usage. Changed name to `updateEnergyUsage`

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts (hunk 5 above)
Comment (vitashubin) [reply biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts]: Can we apply the fix here? Doesn't make sense to push a change that is incorrect even if it is immediately solved in the future.

**Review by vitashubin — APPROVED** [review]
Please fix the date bug that is fixed in PR 85 here, so that we are not even introducing this bug into the master branch at any point. Pending that change, approved

**Pull request merged**